drv/fpga-devices/src/ecp5.rs

```diff
@@ -0,0 +1,453 @@
+        let v = u32::from_be(self.read(c)?);
+
+        match c {
+            Command::ReadId => {
+                ringbuf_entry!(Trace::ReadId(v, Id::from(v)))
+            }
+            Command::ReadStatus => ringbuf_entry!(Trace::ReadStatus(v)),
+            Command::ReadUserCode => ringbuf_entry!(Trace::ReadUserCode(v)),
+            _ => ringbuf_entry!(Trace::Read32(c, v)),
+        }
+
+        Ok(v)
+    }
+
+    /// Read the Status register
+    pub fn status(&self) -> Result<Status, ImplT::Error> {
+        Ok(Status(self.read32(Command::ReadStatus)?))
+    }
+
+    /// Enable ConfigurationMode, allowing access to certain configuration
+    /// command and the bitstream loading process.
+    pub fn enable_configuration_mode(&self) -> Result<(), ImplT::Error> {
+        self.send_command(Command::EnableConfigurationMode)
+    }
+
+    /// Leave ConfigurationMode, disabling access to certaion commands. In
+    /// addition, if a bitstream was loaded, this will transition the device to
+    /// UserMode.
+    pub fn disable_configuration_mode(&self) -> Result<(), ImplT::Error> {
+        self.send_command(Command::DisableConfigurationMode)
+    }
+
+    /// Wait for the device to finish an operation in progress, sleeping for the
+    /// given duration between polling events.
+    pub fn await_not_busy(&self, sleep_ticks: u64) -> Result<(), ImplT::Error> {
+        while self.status()?.busy() {
+            hl::sleep_for(sleep_ticks);
+        }
+        Ok(())
+    }
+
+    /// Wait for the DONE flag to go high.
+    pub fn await_done(&self, sleep_ticks: u64) -> Result<(), ImplT::Error> {
+        while !self.status()?.done() {
+            hl::sleep_for(sleep_ticks);
+        }
+        Ok(())
+    }
+}
+
+impl<ImplT: Ecp5Impl> From<ImplT> for Ecp5<ImplT> {
+    fn from(i: ImplT) -> Self {
+        Ecp5(i)
+    }
+}
+
+pub const DEVICE_RESET_DURATION: u64 = 25;
+pub const APPLICATION_RESET_DURATION: u64 = 25;
+pub const BUSY_DURATION: u64 = 10;
+pub const DONE_DURATION: u64 = 10;
+
+/// Implement the FPGA trait for ECP5, allowing the device to be exposed through
+/// the FPGA server.
+impl<ImplT: Ecp5Impl> Fpga for Ecp5<ImplT>
+where
+    FpgaError: From<<ImplT as Ecp5Impl>::Error>,
+{
+    fn device_enabled(&self) -> Result<bool, FpgaError> {
+        Ok(self.0.program_n()?)
+    }
+
+    fn set_device_enable(&mut self, enabled: bool) -> Result<(), FpgaError> {
+        self.0.set_program_n(enabled)?;
+        ringbuf_entry!(if enabled {
+            Trace::Enabled
+        } else {
+            Trace::Disabled
+        });
+        Ok(())
+    }
+
+    fn reset_device(&mut self, ticks: u64) -> Result<(), FpgaError> {
+        self.set_device_enable(false)?;
+        hl::sleep_for(ticks);
+        self.set_device_enable(true)?;
+        Ok(())
+    }
+
+    fn device_state(&self) -> Result<DeviceState, FpgaError> {
+        if !self.0.program_n()? {
+            Ok(DeviceState::Disabled)
+        } else {
```

Comment:
Clippy will probably complain about this, since you can rewrite it as an `else if` chain